engine/src/main/java/org/terasology/world/chunks/pipeline/LightMergerChunkTaskProvider.java

```diff
@@ -0,0 +1,112 @@
+// Copyright 2020 The Terasology Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+package org.terasology.world.chunks.pipeline;
+
+import com.google.common.collect.Sets;
+import org.jetbrains.annotations.NotNull;
+import org.joml.Vector3i;
+import org.terasology.math.JomlUtil;
+import org.terasology.world.block.BlockRegion;
+import org.terasology.world.block.BlockRegionIterable;
+import org.terasology.world.chunks.Chunk;
+import org.terasology.world.chunks.ChunkProvider;
+import org.terasology.world.chunks.pipeline.tasks.InterruptChunkTask;
+import org.terasology.world.chunks.pipeline.tasks.LightMergerChunkTask;
+import org.terasology.world.propagation.light.LightMerger;
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+/**
+ * Light Merger Chunk Task Provider.
+ * <p>
+ * Check requirements and re-invoke {@link LightMergerChunkTask} when requirements meets.
+ */
+public class LightMergerChunkTaskProvider implements Function<Chunk, ChunkTask>, ChunkTaskListener,
+        ChunkRemoveFromPipelineListener {
+
+    private final ChunkProvider chunkProvider;
+    private final ChunkProcessingPipeline pipeline;
+    private final Set<Chunk> pending = Sets.newConcurrentHashSet();
+    private final Map<Vector3i, Chunk> noticedChunkInProcessing = new ConcurrentHashMap<>();
+
+    public LightMergerChunkTaskProvider(ChunkProvider chunkProvider,
+                                        ChunkProcessingPipeline pipeline) {
+        this.chunkProvider = chunkProvider;
+        this.pipeline = pipeline;
+    }
+
+    @Override
+    public ChunkTask apply(Chunk chunk) {
+        if (getLocalChunks(chunk).noneMatch(Objects::isNull)) {
+            return makeTask(chunk, getLocalChunks(chunk).toArray(Chunk[]::new));
+        } else {
+            pending.add(chunk);
+            return new InterruptChunkTask();
```

Comment:
I am still confused about what the role of InterruptChunkTask is here.

Reply:
Interrupt chunk processing for this `Chunk`.

Kinda "I want to stop processing there, but leave next stages. I will back later."

`LightMerger` requires neighbor chunks for light calculation. then I interrupt chunk processing, while not I not gather required chunks. See `this::onDone`

I had think about `RecursiveTask`, `ForkJoinPool`... but we have `TaskMaster`

Reply:
Ah, I haven't yet studied this to completely understand, but it does help to know this is the reason.

I think this is something it would help to see tests for! Either as tests of Pipeline or tests of LightMergerChunkTaskProvider, whichever seems more appropriate.